File: app/src/main/java/com/ivantha/playtolearn/widget/BoardTileLayout.java
package com.ivantha.playtolearn.widget;

import android.content.Context;
import android.util.TypedValue;
import android.widget.FrameLayout;

import com.ivantha.playtolearn.R;
import com.ivantha.playtolearn.common.Session;
import com.ivantha.playtolearn.listener.BoardDragListener;
import com.ivantha.playtolearn.listener.BoardTouchListener;
import com.ivantha.playtolearn.model.BoardTile;

public class BoardTileLayout extends FrameLayout {
    private BoardTile boardTile;

    public BoardTileLayout(Context context, BoardTile boardTile) {
        super(context);
        this.boardTile = boardTile;

        this.setOnTouchListener(new BoardTouchListener(context));
        this.setOnDragListener(new BoardDragListener());

        applyCustomDesign(context);

        if(boardTile.getBoardTileState() == BoardTile.BoardTileState.CURRENT){
            this.addView(new BoardTileButton(context));
        }
    }

    public BoardTile getBoardTile() {
        return boardTile;
    }

    public void setBoardTileState(BoardTile.BoardTileState boardTileState) {
        this.boardTile.setBoardTileState(boardTileState);
        this.refreshStateChanges();
    }

    public int getRow(){
        return this.boardTile.getRow();
    }

    public int getColumn(){
        return  this.boardTile.getColumn();
    }

    public void generateQuestion(){
//        this.boardTile.setQuestion(QuestionCache.getQuestionArrayList().get(1));
    }

    public int getScore(){
        //Should fix
        return 1;
    }

    public void refreshStateChanges(){
        BoardTile.BoardTileState boardTileState = boardTile.getBoardTileState();
        switch (boardTileState){
            case NOT_VISITED:
                setBackgroundResource(R.drawable.board_tile_background_not_visited0);
//                setBackgroundColor(Color.parseColor("#00ff00"));
                break;
            case VISITED:
                setBackgroundResource(R.drawable.board_tile_background_visited);
                break;
            case CURRENT:
                setBackgroundResource(R.color.colorTransparent);
                break;
            case CORRECT_ANSWER:
                setBackgroundResource(R.drawable.board_tile_background_correct);
                break;
            case WRONG_ANSWER:
                setBackgroundResource(R.drawable.board_tile_background_wrong);
                break;
            default:
                setBackgroundResource(R.drawable.board_tile_background_not_visited0);
//                setBackgroundColor(Color.parseColor("#00ff00"));
                break;
        }
    }

    public void applyCustomDesign(Context context) {
        int outer_px = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 20, context.getResources().getDisplayMetrics());
        int effectiveWidth = Session.SCREEN_WIDTH;

//        int size_px = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 50,
//                context.getResources().getDisplayMetrics());
        int size_px = effectiveWidth / (Session.COLUMN_COUNT + 1);

        //        int margin_px = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 3,
//                context.getResources().getDisplayMetrics());
        int margin_px = (effectiveWidth - (size_px * Session.COLUMN_COUNT)) / (Session.COLUMN_COUNT * 2);

        LayoutParams params = new LayoutParams(size_px, size_px);
        params.setMargins(margin_px, margin_px, margin_px, margin_px);
        this.setLayoutParams(params);

        refreshStateChanges();
    }
}

File: app/src/main/java/com/ivantha/playtolearn/widget/BoardTileLayout.kt
package com.ivantha.playtolearn.widget

import android.content.Context
import android.content.res.Resources
import android.widget.FrameLayout
import com.ivantha.playtolearn.R
import com.ivantha.playtolearn.common.Session
import com.ivantha.playtolearn.listener.BoardDragListener
import com.ivantha.playtolearn.listener.BoardTouchListener
import com.ivantha.playtolearn.model.Tile

class BoardTileLayout(context: Context, val tile: Tile) : FrameLayout(context) {

    //Should fix
    val score: Int
        get() = 1

    init {

        this.setOnTouchListener(BoardTouchListener(context))
        this.setOnDragListener(BoardDragListener())

        applyCustomDesign()

        if (tile.boardTileState === Tile.BoardTileState.CURRENT) {
            this.addView(BoardTileButton(context))
        }
    }

    fun setBoardTileState(tileState: Tile.BoardTileState) {
        this.tile.boardTileState = tileState
        this.refreshStateChanges()
    }

    fun generateQuestion() {
        println("Generate question>>>>>>>>>>>>>>>>>>")
    }

    private fun refreshStateChanges() {
        val boardTileState = tile.boardTileState
        when (boardTileState) {
            Tile.BoardTileState.NOT_VISITED -> setBackgroundResource(R.drawable.board_tile_background_not_visited0)
            Tile.BoardTileState.VISITED -> setBackgroundResource(R.drawable.board_tile_background_visited)
            Tile.BoardTileState.CURRENT -> setBackgroundResource(R.color.colorTransparent)
            Tile.BoardTileState.CORRECT_ANSWER -> setBackgroundResource(R.drawable.board_tile_background_correct)
            Tile.BoardTileState.WRONG_ANSWER -> setBackgroundResource(R.drawable.board_tile_background_wrong)
            else -> setBackgroundResource(R.drawable.board_tile_background_not_visited0)
        }
    }

    private fun applyCustomDesign() {
        val effectiveWidth = Resources.getSystem().displayMetrics.widthPixels
        val sizePx = effectiveWidth / (Session.COLUMN_COUNT + 1)
        val marginPx = (effectiveWidth - sizePx * Session.COLUMN_COUNT) / (Session.COLUMN_COUNT * 2)

        val params = FrameLayout.LayoutParams(sizePx, sizePx)
        params.setMargins(marginPx, marginPx, marginPx, marginPx)
        this.layoutParams = params

        refreshStateChanges()
    }
}
